### PR TITLE
[optimize] provide a better defer operator

### LINKED ISSUE #5576: [optimize] a better defer operater
**Is your feature request related to a problem? Please describe.**

Our current defer operation is based on std:: function But in most C + + standard libraries, when creating a function, there will be a new overhead, which is not necessary for the defer operation. But in fact, if we pass in a lambda expression, we can determine it during compilation, so we can avoid a new / delete overhead during the defer operation

STL std::funcional implement
```c++ 
private:
  static void
  _M_init_functor(_Any_data& __functor, _Functor&& __f, true_type)
  { ::new (__functor._M_access()) _Functor(std::move(__f)); }

  static void
  _M_init_functor(_Any_data& __functor, _Functor&& __f, false_type)
  { __functor._M_access<_Functor*>() = new _Functor(std::move(__f)); }
```

we could use this code instread of DeferOP
```c++
template <class T>
class Defer {
public:
    Defer(T& closure) : _closure(closure) {}
    ~Defer() { _closure(); }

private:
    T& _closure;
};
```

**Describe the solution you'd like**
Here was the bench mark: https://quick-bench.com/q/4IKpxAA5VEbXVziV1G2Po-ZupSE

image

### PR DESCRIPTION
**Pull request by stdpain**
## Proposed changes

will close #5576 

## Types of changes

What types of changes does your code introduce to Doris?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)
- [x] Code refactor (Modify the code structure, format the code, etc...)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have created an issue on (#5576) and described the bug/feature there in detail
- [x] Compiling and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If these changes need document changes, I have updated the document
- [ ] Any dependent changes have been merged

## Further comments

If this is a relatively large or complex change, kick off the discussion at dev@doris.apache.org by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/util/file_utils.cpp [hunk 1]
```diff
@@ -240,16 +240,15 @@ Status FileUtils::copy_file(const std::string& src_path, const std::string& dest
     }
 
     const int64_t BUF_SIZE = 8192;
-    char* buf = new char[BUF_SIZE];
-    DeferOp free_buf(std::bind<void>(std::default_delete<char[]>(), buf));
+    std::unique_ptr<char[]> buf(new char[BUF_SIZE]);
```
Comment (yangzhg) [inline be/src/util/file_utils.cpp:243]: std::unique_ptr<char[]> buf = `make_unique<char[]>(BUF_SIZE);`  may be better

File: be/src/util/defer_op.h [hunk 2]
```diff
@@ -34,6 +34,24 @@ class DeferOp {
...
+// for C++17
+// Defer defer {[]{ call something }};
+//
+// for C++11
+// auto op = [] {};
+// Defer<decltype<op>> (op);
+template <class T>
+class Defer {
+public:
+    Defer(T& closure) : _closure(closure) {}
```
Comment (HappenLee) [inline be/src/util/defer_op.h:47]: may only Defer(T&& closure) : _clousure(std::forward<T>(closure)) is better

Comment (stdpain) [reply be/src/util/defer_op.h:47]: std::forward<T>(closure) will got a compile error
such as:
```
error: cannot bind rvalue reference of type 'main()::<lambda()>&&' to lvalue of type 'main()::<lambda()>'
```

File: be/src/util/file_utils.cpp (hunk 1 above)
Comment (stdpain) [reply be/src/util/file_utils.cpp]: fixed

File: be/src/util/defer_op.h (hunk 2 above)
Comment (HappenLee) [reply be/src/util/defer_op.h:47]: I test the code，do not have the compile error.

File: be/src/util/defer_op.h [hunk 3]
```diff
@@ -34,6 +34,24 @@ class DeferOp {
...
 
+// A Better Defer operator #5576
+// for C++17
+// Defer defer {[]{ call something }};
+//
+// for C++11
+// auto op = [] {};
+// Defer<decltype<op>> (op);
+template <class T>
+class Defer {
```
Comment (HappenLee) [inline be/src/util/defer_op.h:45]: USE Defer replace DeferOP directly, do not create a new class,

**Review by HappenLee — CHANGES_REQUESTED** [review]
little problem

**Comment by acelyc111** [comment]
Seems it is similar with `class ScopedCleanup` in be/src/util/scoped_cleanup.h? why not use it?

**Comment by stdpain** [comment]
@acelyc111 
emm... I think "Defer" has a better semantics than "ScopedCleanup"

**Comment by HappenLee** [comment]
> Seems it is similar with `class ScopedCleanup` in be/src/util/scoped_cleanup.h? why not use it?

it is similar with `class ScopedCleanup`,  but more effectively. we shoule union the code chose only one way to use

**Review by HappenLee — APPROVED** [review]
LGTM

**Review by morningman — APPROVED** [review]
LGTM

**Pull request merged**